druntime/src/core/internal/gc/blkcache.d

```diff
@@ -0,0 +1,213 @@
+/**
+BlkInfo thread-local cache. Used for array appending in the conservative GC to avoid the lock when possible.
+
+Note: this used to be in rt.lifetime, but was moved here to allow GCs to take over array operations.
+*/
+module core.internal.gc.blkcache;
+
+import core.memory;
+
+alias BlkInfo = GC.BlkInfo;
+alias BlkAttr = GC.BlkAttr;
+
+/**
+  cache for the lookup of the block info
+  */
+private enum N_CACHE_BLOCKS=8;
+
+// note this is TLS, so no need to sync.
+BlkInfo *__blkcache_storage;
+
+static if (N_CACHE_BLOCKS==1)
+{
+    version=single_cache;
+}
+else
+{
+    //version=simple_cache; // uncomment to test simple cache strategy
+    //version=random_cache; // uncomment to test random cache strategy
+
+    // ensure N_CACHE_BLOCKS is power of 2.
+    static assert(!((N_CACHE_BLOCKS - 1) & N_CACHE_BLOCKS));
+
+    version (random_cache)
+    {
+        int __nextRndNum = 0;
+    }
+    int __nextBlkIdx;
+}
+
+@property BlkInfo *__blkcache() nothrow
+{
+    if (!__blkcache_storage)
+    {
+        import core.stdc.stdlib;
+        import core.stdc.string;
+        import core.thread.threadbase;
+        // allocate the block cache for the first time
+        immutable size = BlkInfo.sizeof * N_CACHE_BLOCKS;
+        // use noscan, since we do not want to pin any cached allocations.
+        __blkcache_storage = cast(BlkInfo*) GC.calloc(size, BlkAttr.NO_SCAN);
+        ThreadBase.getThis().tlsGCData = __blkcache_storage;
+    }
+    return __blkcache_storage;
+}
+
+/**
+ * Indicates whether an address has been marked by the GC.
+ */
+enum IsMarked : int
+{
+         no, /// Address is not marked.
+        yes, /// Address is marked.
+    unknown, /// Address is not managed by the GC.
+}
+
+alias IsMarkedDg = IsMarked delegate(void* addr) nothrow; /// The isMarked callback function.
+
+// we expect this to be called with the lock in place
+void processGCMarks(void* data, scope IsMarkedDg isMarked) nothrow
+{
+    auto cache = cast(BlkInfo*)data;
```

Comment:
silly question: why not take a `BlkInfo*` as the argument to the function?

Reply:
Hmm, I see that there were two functions with the same name. Disregard this.

Reply:
The GC isn't aware of the type of the `void*`.

I'm not sure exactly how to properly handle this. I kind of want a general "this is custom data the GC can set" in the thread to be used during scanning. For now, this is the only thing. The new GC doesn't use this cache, it doesn't even use BlkInfo. So I don't want to set those details up as a requirement.

In the future, maybe the GC could store other data in which case, then it might make sense for the GC to know this is a BlkInfo pointer (for those GCs who use the blkcache).